File: java/src/main/java/it/giannibombelli/wsc2026/booking/application/usecases/BookingConfirming.java
package it.giannibombelli.wsc2026.booking.application.usecases;

import it.giannibombelli.wsc2026.common.utils.Require;

import it.giannibombelli.wsc2026.booking.domain.booking.Booking;
import it.giannibombelli.wsc2026.booking.domain.events.BookingEvent;
import it.giannibombelli.wsc2026.booking.domain.events.BookingResultEvents;
import it.giannibombelli.wsc2026.booking.domain.ports.BookingRepository;
import it.giannibombelli.wsc2026.common.application.UseCase;
import it.giannibombelli.wsc2026.common.application.events.EventPublisher;

import static it.giannibombelli.wsc2026.booking.application.commands.BookingConfirmationCommands.ConfirmBooking;
import static java.util.Objects.requireNonNull;

public final class BookingConfirming implements UseCase<ConfirmBooking, BookingResultEvents> {
    private final BookingRepository bookingRepository;
    private final EventPublisher<BookingEvent> eventPublisher;

    public BookingConfirming(BookingRepository bookingRepository,
                             EventPublisher<BookingEvent> eventPublisher) {
        this.bookingRepository = requireNonNull(bookingRepository);
        this.eventPublisher = requireNonNull(eventPublisher);
    }

    @Override
    public BookingResultEvents invoke(ConfirmBooking cmd) {
        Require.requireArgument(cmd, "command");

        Booking booking = bookingRepository.findById(cmd.aggregateId())
            .orElseThrow(() -> new IllegalStateException("Booking not found for confirmation: " + cmd.aggregateId()));

        BookingResultEvents result = booking.confirm(cmd.giftCardId(), cmd.amount());

        bookingRepository.save(booking);
        eventPublisher.publish(result);
        return result;
    }
}

File: typescript/src/booking/application/usecases/bookingConfirming.ts
// Caso d'uso per la conferma di una prenotazione.

import type { UseCase } from '@/common/application/usecase.js';
import type { EventPublisher } from '@/common/application/events/eventPublisher.js';
import { Booking } from '../../domain/booking/booking.js';
import { BookingId } from '../../domain/booking/bookingId.js';
import type { BookingEvent } from '../../domain/events/bookingEvent.js';
import type { BookingResultEvent } from '../../domain/events/bookingResultEvents.js';
import type { BookingRepository } from '../../domain/ports/bookingRepository.js';
import type { ConfirmBooking } from '../commands/bookingConfirmationCommands.js';
import { requireArgument } from '@/common/utils/requireArgument.js';
import { requireDependency } from '@/common/utils/requireDependency.js';

export class BookingConfirming implements UseCase<ConfirmBooking, BookingResultEvent> {
  private readonly _bookingRepository: BookingRepository;
  private readonly _eventPublisher: EventPublisher<BookingEvent>;

  constructor(bookingRepository: BookingRepository, eventPublisher: EventPublisher<BookingEvent>) {
    requireDependency(bookingRepository, "bookingRepository");
    requireDependency(eventPublisher, "eventPublisher");
    this._bookingRepository = bookingRepository;
    this._eventPublisher = eventPublisher;
  }

  invoke(command: ConfirmBooking): BookingResultEvent {
    requireArgument(command, 'command');

    const booking = this.findBooking(command.aggregateId);

    const result = booking.confirm(command.amount);

    this._bookingRepository.save(booking);
    this._eventPublisher.publish(result);
    return result;
  }

  private findBooking(id: BookingId): Booking {
    const booking = this._bookingRepository.findById(id);
    if (booking === null) {
      throw new Error(`Booking not found for confirmation: ${id.value}`);
    }
    return booking;
  }
}
